core_impl/src/main/java/io/opencensus/stats/MutableViewData.java

```diff
@@ -206,41 +191,148 @@ public MutableAggregation apply(StdDev arg) {
   @VisibleForTesting
   static AggregationData createAggregationData(MutableAggregation aggregation) {
     return aggregation.match(
-        new Function<MutableSum, AggregationData>() {
-          @Override
-          public AggregationData apply(MutableSum arg) {
-            return SumData.create(arg.getSum());
-          }
-        },
-        new Function<MutableCount, AggregationData>() {
-          @Override
-          public AggregationData apply(MutableCount arg) {
-            return CountData.create(arg.getCount());
-          }
-        },
-        new Function<MutableHistogram, AggregationData>() {
-          @Override
-          public AggregationData apply(MutableHistogram arg) {
-            return HistogramData.create(arg.getBucketCounts());
-          }
-        },
-        new Function<MutableRange, AggregationData>() {
-          @Override
-          public AggregationData apply(MutableRange arg) {
-            return RangeData.create(arg.getMin(), arg.getMax());
-          }
-        },
-        new Function<MutableMean, AggregationData>() {
-          @Override
-          public AggregationData apply(MutableMean arg) {
-            return MeanData.create(arg.getMean());
-          }
-        },
-        new Function<MutableStdDev, AggregationData>() {
-          @Override
-          public AggregationData apply(MutableStdDev arg) {
-            return StdDevData.create(arg.getStdDev());
-          }
-        });
+        CreateSumData.INSTANCE,
+        CreateCountData.INSTANCE,
+        CreateHistogramData.INSTANCE,
+        CreateRangeData.INSTANCE,
+        CreateMeanData.INSTANCE,
+        CreateStdDevData.INSTANCE);
+  }
+  
+  // static inner Function classes
+  
+  private static final class CreateMutableSum implements Function<Sum, MutableAggregation> {
```

Comment:
Thanks for looking into this.  I tried adding `? extends` to the TagKey `match` method to test that it could accept a wider range of arguments.  I changed the signature to

```java
  public abstract <T> T match(
      Function<? super TagKeyString, ? extends T> stringFunction,
      Function<? super TagKeyLong, ? extends T> longFunction,
      Function<? super TagKeyBoolean, ? extends T> booleanFunction,
      Function<? super TagKey, ? extends T> defaultFunction);
```

 Then I was able to write this code that uses it:

```java
  static Number method() {
    TagKey key = TagKeyString.create("key");
    Function<TagKeyString, ? extends Number> f =
        new Function<TagKey.TagKeyString, Double>() {

          @Override
          public Double apply(TagKeyString arg) {
            return 1.0;
          }
        };
    return key.match(
        f,
        Functions.<Number>returnConstant(2),
        Functions.<Long>returnConstant(3L),
        Functions.<Integer>returnConstant(4));
  }
```

It compiled with Java 6 and 8.  The method call shows that `match` can take functions returning the super class (`Number`), subclasses (`Long` and `Integer`), and a bounded wildcard (`? extends Number`).